File: judge-worker/internal/judge/judge.go
package judge

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/leetcode-2.0/judge/internal/models"
	s3client "github.com/leetcode-2.0/judge/internal/s3"
)

type TestCaseCache struct {
	data sync.Map // key: problemSlug, value: []s3client.TestCase
}

func NewCache() *TestCaseCache {
	return &TestCaseCache{}
}

type Judge struct {
	s3            *s3client.S3Client
	cache         *TestCaseCache
	imageRegistry string
}

func New(s3 *s3client.S3Client, cache *TestCaseCache, imageRegistry string) *Judge {
	return &Judge{s3: s3, cache: cache, imageRegistry: imageRegistry}
}

// imageFor resolves the full image name for a language.
// With a registry: "123456.dkr.ecr.../judge:python3"
// Without:         "judge-python3:latest"  (local dev)
func (j *Judge) imageFor(lang string) string {
	if j.imageRegistry != "" {
		return j.imageRegistry + "/judge:" + lang
	}
	return "judge-" + lang + ":latest"
}

// langFilename returns the source filename the runner script expects at /code/
func langFilename(lang string) string {
	switch lang {
	case "java":
		return "Main.java" // Java enforces filename == public class name
	case "python3":
		return "solution.py"
	case "javascript":
		return "solution.js"
	case "cpp":
		return "solution.cpp"
	case "go":
		return "solution.go"
	default:
		return "solution"
	}
}

func (j *Judge) getTestCases(slug, s3Key string) ([]s3client.TestCase, error) {
	if cached, ok := j.cache.data.Load(slug); ok {
		return cached.([]s3client.TestCase), nil
	}

	data, err := j.s3.Download(s3Key)
	if err != nil {
		return nil, fmt.Errorf("s3 download %s: %w", s3Key, err)
	}

	var testCases []s3client.TestCase
	if err := json.Unmarshal(data, &testCases); err != nil {
		return nil, fmt.Errorf("unmarshal test cases: %w", err)
	}

	j.cache.data.Store(slug, testCases)
	return testCases, nil
}

// Run judges a submission. It returns a non-nil error ONLY for transient
// infrastructure failures (S3 unreachable, disk full) — these are NOT the
// submitter's fault and must be retried, so the caller should leave the SQS
// message in the queue (eventually routing it to the DLQ) rather than deleting
// it. A nil error means a legitimate verdict was produced (including a user
// RUNTIME_ERROR), and the result should be published and the message deleted.
func (j *Judge) Run(msg models.SubmissionMessage) (models.JudgeResult, error) {
	result := models.JudgeResult{
		SubmissionID: msg.SubmissionID,
	}

	testCases, err := j.getTestCases(msg.ProblemSlug, msg.TestCasesS3Key)
	if err != nil {
		// Transient: test cases live in S3. Don't blame the user — retry.
		return result, fmt.Errorf("failed to load test cases: %w", err)
	}

	// Write the submission code to a temp dir; mount it read-only into the container.
	// Each submission gets its own dir so concurrent runs don't collide.
	codeDir, err := os.MkdirTemp("", "submission-"+msg.SubmissionID+"-*")
	if err != nil {
		// Transient: local disk problem. Retry on another (or the same) worker.
		return result, fmt.Errorf("failed to create code dir: %w", err)
	}
	defer os.RemoveAll(codeDir)

	codePath := filepath.Join(codeDir, langFilename(msg.Language))
	if err := os.WriteFile(codePath, []byte(msg.Code), 0644); err != nil {
		// Transient: local disk problem. Retry.
		return result, fmt.Errorf("failed to write code: %w", err)
	}

	result.TotalCount = len(testCases)
	result.TestCaseResults = make([]models.TestCaseResult, 0, len(testCases))

	image := j.imageFor(msg.Language)
	overallStatus := "ACCEPTED"
	var maxRuntimeMs int64
	var errorOutput string // stderr from the first failing test case (runtime/compile errors)

	for _, tc := range testCases {
		tcResult, status, stderr := runTestCase(image, codeDir, tc, msg)
		result.TestCaseResults = append(result.TestCaseResults, tcResult)

		if tcResult.RuntimeMs > maxRuntimeMs {
			maxRuntimeMs = tcResult.RuntimeMs
		}
		if status != "ACCEPTED" && overallStatus == "ACCEPTED" {
			overallStatus = status
			errorOutput = stderr
		}
		if tcResult.Passed {
			result.PassedCount++
		}
	}

	result.Status = overallStatus
	result.RuntimeMs = maxRuntimeMs

	// Surface the captured stderr so the client can show the actual error.
	if errorOutput != "" {
		const maxLen = 4000
		if len(errorOutput) > maxLen {
			errorOutput = errorOutput[:maxLen] + "\n...(truncated)"
		}
		result.CompileError = errorOutput
	}

	return result, nil
}

// runTestCase returns the per-case result, the verdict status, and (for error
// verdicts) the program's stderr so the caller can surface it to the client.
func runTestCase(image, codeDir string, tc s3client.TestCase, msg models.SubmissionMessage) (models.TestCaseResult, string, string) {
	// wall-clock timeout = problem limit + 2s grace
	timeout := time.Duration(msg.TimeLimitMs+2000) * time.Millisecond
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	args := []string{
		"run", "--rm", "-i",
		"--network", "none",
		fmt.Sprintf("--memory=%dm", msg.MemoryLimitMb),
		"--cpus=0.5",
		"--read-only",
		"--tmpfs", "/tmp",
		// mount the submission code read-only; the run.sh inside each image reads from /code/
		"-v", codeDir + ":/code:ro",
		image,
	}

	cmd := exec.CommandContext(ctx, "docker", args...)
	cmd.Stdin = strings.NewReader(tc.Input) // test case input → program's stdin

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	start := time.Now()
	runErr := cmd.Run()
	runtimeMs := time.Since(start).Milliseconds()

	tcResult := models.TestCaseResult{
		ID:        tc.ID,
		RuntimeMs: runtimeMs,
	}

	if ctx.Err() == context.DeadlineExceeded {
		return tcResult, "TIME_LIMIT_EXCEEDED", ""
	}

	if runErr != nil {
		if exitErr, ok := runErr.(*exec.ExitError); ok {
			// OOM kill — Linux sends SIGKILL → exit 137 (128 + 9)
			if exitErr.ExitCode() == 137 {
				return tcResult, "MEMORY_LIMIT_EXCEEDED", ""
			}
		}
		errOutput := strings.TrimSpace(stderr.String())
		if errOutput != "" {
			log.Printf("submission %s tc %d stderr: %s", msg.SubmissionID, tc.ID, errOutput)
		}
		return tcResult, "RUNTIME_ERROR", errOutput
	}

	if strings.TrimSpace(stdout.String()) == strings.TrimSpace(tc.ExpectedOutput) {
		tcResult.Passed = true
		return tcResult, "ACCEPTED", ""
	}

	return tcResult, "WRONG_ANSWER", ""
}

File: judge-worker/judge.js
import { spawn } from "node:child_process";
import { mkdtemp, writeFile, rm } from "node:fs/promises";
import { join } from "node:path";
import { tmpdir } from "node:os";

/**
 * @typedef {Object} SubmissionMessage
 * @property {string} submissionId
 * @property {string} problemId
 * @property {string} problemSlug
 * @property {string} userId
 * @property {string} language
 * @property {string} code
 * @property {number} timeLimitMs
 * @property {number} memoryLimitMb
 * @property {string} testCasesS3Key
 */

/**
 * @typedef {Object} TestCaseResult
 * @property {number} id
 * @property {boolean} passed
 * @property {number} runtimeMs
 */

/**
 * @typedef {Object} JudgeResult
 * @property {string} submissionId
 * @property {string} status
 * @property {number} passedCount
 * @property {number} totalCount
 * @property {number} runtimeMs
 * @property {number} memoryUsedMb
 * @property {TestCaseResult[]} testCaseResults
 * @property {string} [compileError]
 */

/**
 * @param {string} lang
 * @returns {string}
 */
function langFilename(lang) {
  switch (lang) {
    case "java":
      return "Main.java";
    case "python3":
      return "solution.py";
    case "javascript":
      return "solution.js";
    case "cpp":
      return "solution.cpp";
    case "go":
      return "solution.go";
    default:
      return "solution";
  }
}

export class Judge {
  /**
   * @param {import("./s3.js").TestCasesS3} s3
   * @param {string} imageRegistry
   */
  constructor(s3, imageRegistry) {
    this.s3 = s3;
    this.imageRegistry = imageRegistry;
    /** @type {Map<string, import("./s3.js").TestCase[]>} */
    this.cache = new Map();
  }

  /** @param {string} lang */
  imageFor(lang) {
    if (this.imageRegistry) {
      return `${this.imageRegistry}/judge:${lang}`;
    }
    return `judge-${lang}:latest`;
  }

  /**
   * @param {string} slug
   * @param {string} s3Key
   * @returns {Promise<import("./s3.js").TestCase[]>}
   */
  async getTestCases(slug, s3Key) {
    const cached = this.cache.get(slug);
    if (cached) return cached;

    const data = await this.s3.download(s3Key);
    /** @type {import("./s3.js").TestCase[]} */
    const testCases = JSON.parse(data.toString("utf8"));
    this.cache.set(slug, testCases);
    return testCases;
  }

  /**
   * Judges a submission. Throws ONLY for transient infrastructure failures
   * (S3, disk). A resolved result — including user RUNTIME_ERROR — is a
   * legitimate verdict and should be published + the SQS message deleted.
   *
   * @param {SubmissionMessage} msg
   * @returns {Promise<JudgeResult>}
   */
  async run(msg) {
    /** @type {JudgeResult} */
    const result = {
      submissionId: msg.submissionId,
      status: "ACCEPTED",
      passedCount: 0,
      totalCount: 0,
      runtimeMs: 0,
      memoryUsedMb: 0,
      testCaseResults: [],
    };

    let testCases;
    try {
      testCases = await this.getTestCases(msg.problemSlug, msg.testCasesS3Key);
    } catch (err) {
      throw new Error(`failed to load test cases: ${err.message}`);
    }

    const codeDir = await mkdtemp(join(tmpdir(), `submission-${msg.submissionId}-`));
    try {
      const codePath = join(codeDir, langFilename(msg.language));
      try {
        await writeFile(codePath, msg.code, { mode: 0o644 });
      } catch (err) {
        throw new Error(`failed to write code: ${err.message}`);
      }

      result.totalCount = testCases.length;

      const image = this.imageFor(msg.language);
      let overallStatus = "ACCEPTED";
      let maxRuntimeMs = 0;
      let errorOutput = "";

      for (const tc of testCases) {
        const { tcResult, status, stderr } = await runTestCase(
          image,
          codeDir,
          tc,
          msg
        );
        result.testCaseResults.push(tcResult);

        if (tcResult.runtimeMs > maxRuntimeMs) {
          maxRuntimeMs = tcResult.runtimeMs;
        }
        if (status !== "ACCEPTED" && overallStatus === "ACCEPTED") {
          overallStatus = status;
          errorOutput = stderr;
        }
        if (tcResult.passed) {
          result.passedCount++;
        }
      }

      result.status = overallStatus;
      result.runtimeMs = maxRuntimeMs;

      if (errorOutput) {
        const maxLen = 4000;
        result.compileError =
          errorOutput.length > maxLen
            ? errorOutput.slice(0, maxLen) + "\n...(truncated)"
            : errorOutput;
      }

      return result;
    } finally {
      await rm(codeDir, { recursive: true, force: true });
    }
  }
}

/**
 * @param {string} image
 * @param {string} codeDir
 * @param {import("./s3.js").TestCase} tc
 * @param {SubmissionMessage} msg
 * @returns {Promise<{ tcResult: TestCaseResult, status: string, stderr: string }>}
 */
function runTestCase(image, codeDir, tc, msg) {
  const timeoutMs = msg.timeLimitMs + 2000;

  const args = [
    "run",
    "--rm",
    "-i",
    "--network",
    "none",
    `--memory=${msg.memoryLimitMb}m`,
    "--cpus=0.5",
    "--read-only",
    "--tmpfs",
    "/tmp",
    "-v",
    `${codeDir}:/code:ro`,
    image,
  ];

  return new Promise((resolve) => {
    const start = Date.now();
    const child = spawn("docker", args, { stdio: ["pipe", "pipe", "pipe"] });

    let stdout = "";
    let stderr = "";
    let timedOut = false;
    let settled = false;

    const timer = setTimeout(() => {
      timedOut = true;
      child.kill("SIGKILL");
    }, timeoutMs);

    child.stdout.on("data", (chunk) => {
      stdout += chunk.toString();
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk.toString();
    });

    child.stdin.write(tc.input);
    child.stdin.end();

    child.on("error", (err) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      const runtimeMs = Date.now() - start;
      console.error(
        `submission ${msg.submissionId} tc ${tc.id} docker spawn error: ${err.message}`
      );
      resolve({
        tcResult: { id: tc.id, passed: false, runtimeMs },
        status: "RUNTIME_ERROR",
        stderr: err.message,
      });
    });

    child.on("close", (code) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      const runtimeMs = Date.now() - start;

      /** @type {TestCaseResult} */
      const tcResult = { id: tc.id, passed: false, runtimeMs };

      if (timedOut) {
        resolve({ tcResult, status: "TIME_LIMIT_EXCEEDED", stderr: "" });
        return;
      }

      // OOM kill — Linux SIGKILL → exit 137 (128 + 9)
      if (code === 137) {
        resolve({ tcResult, status: "MEMORY_LIMIT_EXCEEDED", stderr: "" });
        return;
      }

      if (code !== 0) {
        const errOutput = stderr.trim();
        if (errOutput) {
          console.log(
            `submission ${msg.submissionId} tc ${tc.id} stderr: ${errOutput}`
          );
        }
        resolve({ tcResult, status: "RUNTIME_ERROR", stderr: errOutput });
        return;
      }

      if (stdout.trim() === tc.expectedOutput.trim()) {
        tcResult.passed = true;
        resolve({ tcResult, status: "ACCEPTED", stderr: "" });
        return;
      }

      resolve({ tcResult, status: "WRONG_ANSWER", stderr: "" });
    });
  });
}
